### PR TITLE
[JIT-1016] Add metrics to unbounded queues + misc fixes

### PR DESCRIPTION
**Pull request by esemeniuc**
Notes:
- Clippy fails due to missing protoc, fixed in https://github.com/jito-foundation/jito-relayer/pull/72

Changes:
- Log unbounded & large channel sizes
  - Convert to bounded channels after giving about 10x margin to peak usage
- Handle partially exited threads (use panic)
- Fix incorrect Dashmap cloning
- Remove unnecessary mutexes
- Backport ./b simplification
- Backport Dockerfile caching improvements
- Library updates
- Clap parsing cleanups
- Fix docker-compose start (missing params)
- Clean up redundant functions

### CODE REVIEW

File: core/src/fetch_stage.rs [hunk 1]
```diff
@@ -8,40 +8,72 @@ use std::{
...
 
 #[derive(Debug, thiserror::Error)]
-pub enum Error {
-    #[error("send error")]
-    Send,
+pub enum FetchStageError {
+    #[error("send error: {0}")]
+    Send(#[from] SendError<PacketBatch>),
     #[error("recv timeout")]
     RecvTimeout(#[from] RecvTimeoutError),
```
Comment (buffalu) [inline core/src/fetch_stage.rs:22]: some of this is kinda jank, can prob clean up these erorrsr (thats on me lol)

File: transaction-relayer/src/forwarder.rs [hunk 2]
```diff
@@ -165,26 +165,36 @@ pub fn start_forward_and_delay_thread(
...
+                                .send(batch)
+                                .expect("exiting forwarding delayed packets");
+                        }
+
+                        if iter_count % CHANNEL_REPORT_INTERVAL == 0 {
+                            datapoint_info!(
+                                "forwarder-channel_stats",
+                                (
+                                    "buffered_packet_batches-queue_len",
+                                    buffered_packet_batches.len(),
```
Comment (buffalu) [inline transaction-relayer/src/forwarder.rs:192]: can you call buffered_packet_batches.capacity() here?

File: relayer/src/health_manager.rs [hunk 3]
```diff
@@ -80,18 +80,19 @@ impl HealthManager {
...
+                            let slot = maybe_slot.expect("error receiving slot, exiting");
+                            slot_sender.send(slot).expect("error forwarding slot, exiting");
                             last_update = Instant::now();
                         }
                     }
+
+                    if iter_count % HealthManager::CHANNEL_REPORT_INTERVAL == 0 {
+                        datapoint_info!(
+                                    "health_manager-channel_stats",
+                                    ("slot_sender_queue_len", slot_sender.len(), i64),
```
[the remark this answers is not in the document]
Comment (esemeniuc) [reply relayer/src/health_manager.rs]: updated to max

File: b [hunk 4]
```diff
@@ -1,40 +1,23 @@
...
-    RPC_SERVERS=http://172.17.0.1:8899
-    WEBSOCKET_SERVERS=ws://172.17.0.1:8900
-    BLOCK_ENGINE_AUTH_SERVICE_URL=http://172.17.0.1:1005
-    BLOCK_ENGINE_URL=http://172.17.0.1:1000
-  else
-    echo "unsupported testing platform, exiting"
-    exit 1
-  fi
+export DOCKER_LOCALHOST=172.17.0.1
+if [[ $(uname) == "Darwin" ]]; then
```
Comment (segfaultdoc) [inline b:8]: wanna add back the else that exits if unsupported platform?

File: block_engine/src/block_engine.rs [hunk 5]
```diff
@@ -167,8 +132,6 @@ impl BlockEngineRelayerHandler {
                                 datapoint_error!("block_engine_relayer-error",
                                     "block_engine_url" => block_engine_url,
                                     "auth_service_url" => auth_service_url,
-                                    "cluster" => &cluster,
```
Comment (segfaultdoc) [inline block_engine/src/block_engine.rs:170]: why remove?

Comment (esemeniuc) [reply block_engine/src/block_engine.rs:170]: its part of the host_id now, no need to manually pass it around

File: block_engine/src/block_engine.rs [hunk 6]
```diff
@@ -475,11 +436,17 @@ impl BlockEngineRelayerHandler {
...
+                    block_engine_stats.increment_programs_of_interest_len(programs_of_interest.cache_size() as u64);
 
-                    block_engine_stats.report(&cluster, &region);
+                    block_engine_stats.report();
                     block_engine_stats = BlockEngineStats::default();
                 }
             }
+
+            block_engine_stats.update_block_engine_packet_sender_len(
+                (Self::BLOCK_ENGINE_PACKET_QUEUE_CAPACITY - block_engine_packet_sender.capacity())
```
Comment (segfaultdoc) [inline block_engine/src/block_engine.rs:447]: I think emitting capacity would be useful and easy to tell if the queue is filling up without looking at the code

File: b (hunk 4 above)
Comment (esemeniuc) [reply b:8]: what other platforms would you like to guard against? windows cant run bash scripts without wsl, and wsl presents itself as linux. this is a direct copy from block engine

File: block_engine/src/block_engine.rs (hunk 5 above)
Comment (esemeniuc) [reply block_engine/src/block_engine.rs:170]: there were a bunch of metrics that didnt have cluster and region, grafana couldnt filter well

File: b (hunk 4 above)
Comment (segfaultdoc) [reply b:8]: I'm just referring to this: https://github.com/jito-foundation/jito-relayer/pull/69/files#diff-3e23e8160039594a33894f6564e1b1348bbd7a0088d42c4acb73eeaed59c009dL17

File: core/src/fetch_stage.rs [hunk 7]
```diff
@@ -6,53 +6,99 @@ use std::{
...
+        let fwd_thread_hdl = Builder::new()
+            .name("fetch_stage-forwarder_thread".to_string())
+            .spawn(move || {
+                let metrics_interval = Duration::from_secs(1);
+                let mut start = Instant::now();
+                let mut tpu_forwards_receiver_max_len = 0usize;
+                let mut tpu_sender_max_len = 0usize;
+                while !exit.load(Ordering::Relaxed) {
+                    match Self::handle_forwarded_packets(&tpu_forwards_receiver, &tpu_sender) {
+                        Ok(()) | Err(FetchStageError::RecvTimeout(RecvTimeoutError::Timeout)) => {}
```
Comment (segfaultdoc) [inline core/src/fetch_stage.rs:49]: looks like fn never returns a timeout, wdyt we rm this from the Error enum type since nothing returns that error here?

Comment (segfaultdoc) [reply core/src/fetch_stage.rs:49]: alternatively should we add a timeout equal to metrics_interval so we don't have no-data on the below metrics when no packets are coming thru?

File: core/src/staked_nodes_updater_service.rs [hunk 8]
```diff
@@ -59,10 +59,10 @@ impl StakedNodesUpdaterService {
         ip_to_stake: &mut HashMap<IpAddr, u64>,
         pubkey_stake_map: &mut HashMap<Pubkey, u64>,
         total_stake: &mut u64,
-        rpc_load_balancer: &Arc<Mutex<LoadBalancer>>,
+        rpc_load_balancer: &Arc<LoadBalancer>,
```
Comment (segfaultdoc) [inline core/src/staked_nodes_updater_service.rs:62]: naishe

**Review by segfaultdoc — APPROVED** [review]

**Review by segfaultdoc — COMMENTED** [review]
lgtm for future PRs can you break down to be more digestable, and have separate diffs for changes that aren't explicitly related e.g. package upgrades

File: b (hunk 4 above)
Comment (esemeniuc) [reply b:13]: added check

**Pull request merged**